File: recursive_sum_product_array.rb
# http://rosettacode.org/wiki/Sum_and_product_of_an_array

def sum(array,index=0)
  index >= array.count ? 0 : array[index] + sum(array,index+1)
end

def product(array,index=0)
  index >= array.count ? 1 : array[index] * product(array,index+1)
end

test_array_one = [0,1,2,3]
test_array_two = [1,2,3,4]

p sum(test_array_one)
p sum(test_array_two)
p product(test_array_one)
p product(test_array_two)

File: recursive_sum_product_array.swift
// http://rosettacode.org/wiki/Sum_and_product_of_an_array

func sum(array:[Int],index:Int) -> Int {
  if index >= array.count {
    return 0
  } else {
    return array[index] + sum(array,index+1)
  }
}

func product(array:[Int],index:Int) -> Int {
  if index >= array.count {
    return 1
  } else {
    return array[index] * product(array,index+1)
  }
}

let testArrayOne = [0,1,2,3]
let testArrayTwo = [1,2,3,4]

println(sum(testArrayOne,0))
println(sum(testArrayTwo,0))
println(product(testArrayOne,0))
println(product(testArrayTwo,0))
